### PR TITLE
Issue/10

### LINKED ISSUE #10: Port asset functions
- [x] _totalAssets
- [x] totalAssets
- [x] _calculateLockedProfit
- [x] _freeFunds
- [x] _reportLoss

### PR DESCRIPTION
**Pull request by bitcoinbrisbane**
Closes #10

### CODE REVIEW

**Review by 0x1NotMe — COMMENTED** [review]
LGTM

File: contracts/Vault.sol [hunk 1]
```diff
@@ -351,7 +362,22 @@ contract Vault is ERC20 {
     }
 
     function _reportLoss(address strategy, uint256 loss) internal {
-        
+        uint256 totalDebt = strategies[strategy].totalDebt;
+        assert(totalDebt >= loss);
+
+        if (debtRatio != 0) {
+            // todo: min
```
Comment (0x1NotMe) [inline contracts/Vault.sol:369]: Is this still a relevant todo?

**Review by 0x1NotMe — APPROVED** [review]

File: contracts/Vault.sol (hunk 1 above)
Comment (bitcoinbrisbane) [reply contracts/Vault.sol:369]: Nope, but ill remove the comment.

File: contracts/Vault.sol [hunk 2]
```diff
@@ -351,7 +362,21 @@ contract Vault is ERC20 {
     }
 
     function _reportLoss(address strategy, uint256 loss) internal {
-        
+        uint256 totalDebt = strategies[strategy].totalDebt;
```
Comment (LichuAcu) [inline contracts/Vault.sol:365]:
```suggestion
        uint256 _totalDebt = strategies[strategy].totalDebt;
```

File: contracts/Vault.sol [hunk 3]
```diff
@@ -351,7 +362,21 @@ contract Vault is ERC20 {
...
+        assert(totalDebt >= loss);
```
Comment (LichuAcu) [inline contracts/Vault.sol:366]:
```suggestion
        assert(_totalDebt >= loss);
```

File: contracts/Vault.sol [hunk 4]
```diff
@@ -351,7 +362,21 @@ contract Vault is ERC20 {
...
+            uint256 ratio_change = Math.min(loss * debtRatio / totalDebt, strategies[strategy].debtRatio);
+
+            if (ratio_change != 0) {
+                strategies[strategy].debtRatio -= ratio_change;
+                debtRatio -= ratio_change;
+            }
+        }
+
+        strategies[strategy].totalLoss += loss;
+        strategies[strategy].totalDebt = totalDebt - loss;
```
Comment (LichuAcu) [inline contracts/Vault.sol:378]:
``` suggestion
        strategies[strategy].totalDebt = _totalDebt - loss;
```

File: contracts/Vault.sol [hunk 5]
```diff
@@ -351,7 +362,21 @@ contract Vault is ERC20 {
     }
 
     function _reportLoss(address strategy, uint256 loss) internal {
-        
+        uint256 totalDebt = strategies[strategy].totalDebt;
+        assert(totalDebt >= loss);
+
+        if (debtRatio != 0) {
+            uint256 ratio_change = Math.min(loss * debtRatio / totalDebt, strategies[strategy].debtRatio);
```
Comment (LichuAcu) [inline contracts/Vault.sol:369]: totalDebt should not be changed here, since in the Vyper version they use self.totalDebt, which is the state variable

**Review by LichuAcu — COMMENTED** [review]
Solved some issues regarding _totalDebt_ shadowed declaration.

**Pull request merged**